batch/batch/batch.py

```diff
@@ -93,51 +91,59 @@ def batch_record_to_dict(record):
-        db,
-        'CALL mark_job_complete(%s, %s, %s, %s, %s, %s, %s, %s, %s);',
-        (batch_id, job_id, attempt_id, new_state,
-         json.dumps(status) if status is not None else None,
-         start_time, end_time, reason, now))
+    try:
+        rv = await db.execute_and_fetchone(
+            'CALL mark_job_complete(%s, %s, %s, %s, %s, %s, %s, %s, %s, %s);',
+            (batch_id, job_id, attempt_id, instance_name, new_state,
+             json.dumps(status) if status is not None else None,
+             start_time, end_time, reason, now))
+    except:
+        log.exception(f'error while marking job {id} complete on instance {instance_name}')
+        raise
 
-    log.info(f'mark_job_complete returned {rv} for job {id}')
+    if instance_name:
+        instance = inst_pool.name_instance.get(instance_name)
+        if instance:
+            if rv['delta_cores_mcpu'] != 0 and instance.state == 'active':
+                instance.adjust_free_cores_in_memory(rv['delta_cores_mcpu'])
+                scheduler_state_changed.set()
+        else:
+            log.warning(f'mark_complete for job {id} from unknown {instance}')
+
+    if rv['rc'] != 0:
+        log.info(f'mark_job_complete returned {rv} for job {id}')
+        return
 
     old_state = rv['old_state']
     if old_state in complete_states:
-        log.info(f'old_state {old_state} complete, doing nothing')
+        log.info(f'old_state {old_state} complete for job {id}, doing nothing')
         # already complete, do nothing
         return
 
     log.info(f'job {id} changed state: {rv["old_state"]} => {new_state}')
 
-    instance_name = rv['instance_name']
-    if instance_name:
-        instance = inst_pool.name_instance.get(instance_name)
-        if instance:
-            log.info(f'updating {instance}')
-
-            instance.adjust_free_cores_in_memory(rv['cores_mcpu'])
-            scheduler_state_changed.set()
-        else:
-            log.warning(f'mark_complete for job {id} from unknown {instance}')
-
     await notify_batch_job_complete(db, batch_id)
 
 
-async def mark_job_started(app, batch_id, job_id, attempt_id, start_time):
+async def mark_job_started(app, batch_id, job_id, attempt_id, instance, start_time):
     db = app['db']
 
     id = (batch_id, job_id)
 
     log.info(f'mark job {id} started')
 
-    await db.execute_update(
-        '''
-UPDATE attempts
-SET start_time = %s
-WHERE batch_id = %s AND job_id = %s AND attempt_id = %s;
-''',
-        (start_time, batch_id, job_id, attempt_id))
+    try:
+        rv = await db.execute_and_fetchone(
+            '''
+    CALL mark_job_started(%s, %s, %s, %s, %s);
+    ''',
+            (batch_id, job_id, attempt_id, instance.name, start_time))
+    except:
+        log.exception(f'error while marking job {id} started on {instance}')
+        raise
+
+    if rv['delta_cores_mcpu'] != 0 and instance.state == 'active':
+        instance.adjust_free_cores_in_memory(rv['delta_cores_mcpu'])
```

Comment:
I think I assumed a mark job started would always remove free cores or do nothing to the free cores, so it didn't make sense to have a new scheduling loop in that case because there's no addition of free cores.